airflow/jobs/backfill_job.py

```diff
@@ -224,10 +242,19 @@ def _update_counters(self, ti_status):
                     "reaching concurrency limits. Re-adding task to queue.",
                     ti
                 )
-                ti.set_state(State.SCHEDULED)
+                tis_to_be_scheduled.append(ti)
                 ti_status.running.pop(key)
                 ti_status.to_run[key] = ti
 
+        # Batch schedule of task instances
+        if tis_to_be_scheduled:
+            filter_for_tis = ([and_(TI.dag_id == ti.dag_id,
+                                    TI.task_id == ti.task_id,
+                                    TI.execution_date == ti.execution_date)
+                               for ti in tis_to_be_scheduled])
```

Comment:
I think we can extract this to `TaskInstance.filter_by_tis`. My concern is: should we support both `Iterbale[TaskInstance]` and `Iterable[TaskInstanceKey]`?